### ISSUE TITLE
sprintf vector flag flips parameter order

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11109]
Migrated from [rt.perl.org#83194](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=83194) (status was 'resolved')

Searchable as RT83194$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



\#\!perl
use warnings;
use strict;

use Test&#8203;::More tests => 8;

is sprintf\('%\.3X'\, 11\)\, '00B'\, 'hex\, zero\-padded to 3 places';
is sprintf\('%\.\*X'\, 3\, 11\)\, '00B'\, 'dynamic precision';
is sprintf\('%vX'\, '012'\)\, '30\.31\.32'\, 'vector flag';
is sprintf\('%\*vX'\, '&#8203;:'\, '012'\)\, '30&#8203;:31&#8203;:32'\, 'vector flag \+ custom separator';
is sprintf\('%v\.3X'\, '012'\)\, '030\.031\.032'\, 'vector flag \+ static precision';
is sprintf\('%v\.\*X'\, 3\, '012'\)\, '030\.031\.032'\, 'vector flag \+ dynamic precision';
is sprintf\('%\*v\.3X'\, '&#8203;:'\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+ custom separator \+ static precision';
is sprintf\('%\*v\.\*X'\, '&#8203;:'\, 3\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+ custom separator \+ dynamic precision';

\_\_END\_\_

1\.\.8                                                                                  
ok 1 \- hex\, zero\-padded to 3 places
ok 2 \- dynamic precision
ok 3 \- vector flag
ok 4 \- vector flag \+ custom separator
ok 5 \- vector flag \+ static precision
not ok 6 \- vector flag \+ dynamic precision
\#   Failed test 'vector flag \+ dynamic precision'
\#   at /tmp/fBQVePzsV7 line 12\.
\#          got&#8203;: '000000000033'
\#     expected&#8203;: '030\.031\.032'
ok 7 \- vector flag \+ custom separator \+ static precision
not ok 8 \- vector flag \+ custom separator \+ dynamic precision
\#   Failed test 'vector flag \+ custom separator \+ dynamic precision'
\#   at /tmp/fBQVePzsV7 line 14\.
\#          got&#8203;: '000000000033'
\#     expected&#8203;: '030&#8203;:031&#8203;:032'
\# Looks like you failed 2 tests of 8\.


For some reason combining the vector flag and dynamic precision causes
sprintf to confuse the order of its arguments\, making it treat the string
as the precision and vice versa \('000000000033' is sprintf '%v12X'\, '3'\)\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl 5.12.1 - Thu Jun  3 20:09:15 CEST 2010
It is being executed now by  Perl 5.12.2 - Thu Jan  6 22:40:00 CET 2011.

Site configuration information for perl 5.12.2:

Configured by mauke at Thu Jan  6 22:40:00 CET 2011.

Summary of my perl5 (revision 5 version 12 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.27-gentoo-r8, archname=i686-linux
    uname='linux nora 2.6.27-gentoo-r8 #4 preempt mon apr 13 21:59:34 cest 2009 i686 amd athlon(tm) 64 processor 3200+ authenticamd gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -march=native',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.5.1', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -march=native -L/usr/local/lib -fstack-protector'

Locally applied patches:
    SAVEARGV0 - disable magic open in <ARGV>


@INC for perl 5.12.2:
    /home/mauke/usr/local/lib/perl5/site_perl/5.12.2/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.12.2
    /home/mauke/usr/local/lib/perl5/5.12.2/i686-linux
    /home/mauke/usr/local/lib/perl5/5.12.2
    /home/mauke/usr/local/lib/perl5/site_perl/5.12.1/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.12.1
    /home/mauke/usr/local/lib/perl5/site_perl
    .


Environment for perl 5.12.2:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/local/lib
    LOGDIR (unset)
    PATH=/home/mauke/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.4.3:/opt/sun-jdk-1.4.2.13/bin:/opt/sun-jdk-1.4.2.13/jre/bin:/opt/sun-jdk-1.4.2.13/jre/javaws:/opt/dmd/bin:/usr/games/bin:/opt/vmware/server/bin
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @hvds

"l\.mai@&#8203;web\.de \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
\[\.\.\.\]
:is sprintf\('%\.3X'\, 11\)\, '00B'\, 'hex\, zero\-padded to 3 places';
:is sprintf\('%\.\*X'\, 3\, 11\)\, '00B'\, 'dynamic precision';
:is sprintf\('%vX'\, '012'\)\, '30\.31\.32'\, 'vector flag';
:is sprintf\('%\*vX'\, '&#8203;:'\, '012'\)\, '30&#8203;:31&#8203;:32'\, 'vector flag \+ custom separator';
:is sprintf\('%v\.3X'\, '012'\)\, '030\.031\.032'\, 'vector flag \+ static precision';
:is sprintf\('%v\.\*X'\, 3\, '012'\)\, '030\.031\.032'\, 'vector flag \+ dynamic precision';
:is sprintf\('%\*v\.3X'\, '&#8203;:'\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+ custom separator \+ static precision';
:is sprintf\('%\*v\.\*X'\, '&#8203;:'\, 3\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+ custom separator \+ dynamic precision';
:
:\_\_END\_\_
:
:1\.\.8                                                                                  
:ok 1 \- hex\, zero\-padded to 3 places
:ok 2 \- dynamic precision
:ok 3 \- vector flag
:ok 4 \- vector flag \+ custom separator
:ok 5 \- vector flag \+ static precision
:not ok 6 \- vector flag \+ dynamic precision
:\#   Failed test 'vector flag \+ dynamic precision'
:\#   at /tmp/fBQVePzsV7 line 12\.
:\#          got&#8203;: '000000000033'
:\#     expected&#8203;: '030\.031\.032'
:ok 7 \- vector flag \+ custom separator \+ static precision
:not ok 8 \- vector flag \+ custom separator \+ dynamic precision
:\#   Failed test 'vector flag \+ custom separator \+ dynamic precision'
:\#   at /tmp/fBQVePzsV7 line 14\.
:\#          got&#8203;: '000000000033'
:\#     expected&#8203;: '030&#8203;:031&#8203;:032'
:\# Looks like you failed 2 tests of 8\.

Note that we already have tests expecting the current behaviour \- see
tests 157\-164 in t/op/sprintf\.t\, where the details of test 157 appear
at the 157th line after the \_\_END\_\_ token\.

That doesn't necessarily mean it's correct though\.

My reading of \`perldoc \-f sprintf\` is that your interpretation is correct\,
but I wouldn't be surprised if I wrote both that and the tests\, so I don't
particularly trust my judgment on this\.

Do we have an independent declaration of what it should do from a C\-library
sprintf doc somewhere?

Hugo

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

On Thu\, Feb 3\, 2011 at 8&#8203;:35 AM\,  \<hv@&#8203;crypt\.org> wrote&#8203;:
> Do we have an independent declaration of what it should do from a C\-library
> sprintf doc somewhere?

C versions of sprintf do not have the vector \(v\) flag\.

Ambrus

**Comment by p5pRT** [comment]
### From @hvds

I wrote&#8203;:
:"l\.mai@&#8203;web\.de \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
:\[\.\.\.\]
:&#8203;:is sprintf\('%\.3X'\, 11\)\, '00B'\, 'hex\, zero\-padded to 3 places';
:&#8203;:is sprintf\('%\.\*X'\, 3\, 11\)\, '00B'\, 'dynamic precision';
:&#8203;:is sprintf\('%vX'\, '012'\)\, '30\.31\.32'\, 'vector flag';
:&#8203;:is sprintf\('%\*vX'\, '&#8203;:'\, '012'\)\, '30&#8203;:31&#8203;:32'\, 'vector flag \+ custom separator';
:&#8203;:is sprintf\('%v\.3X'\, '012'\)\, '030\.031\.032'\, 'vector flag \+ static precision';
:&#8203;:is sprintf\('%v\.\*X'\, 3\, '012'\)\, '030\.031\.032'\, 'vector flag \+ dynamic precision';
:&#8203;:is sprintf\('%\*v\.3X'\, '&#8203;:'\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+ custom separator \+ static precision';
:&#8203;:is sprintf\('%\*v\.\*X'\, '&#8203;:'\, 3\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+ custom separator \+ dynamic precision';
:&#8203;:
:&#8203;:\_\_END\_\_
:&#8203;:
:&#8203;:1\.\.8
:&#8203;:ok 1 \- hex\, zero\-padded to 3 places
:&#8203;:ok 2 \- dynamic precision
:&#8203;:ok 3 \- vector flag
:&#8203;:ok 4 \- vector flag \+ custom separator
:&#8203;:ok 5 \- vector flag \+ static precision
:&#8203;:not ok 6 \- vector flag \+ dynamic precision
:&#8203;:\#   Failed test 'vector flag \+ dynamic precision'
:&#8203;:\#   at /tmp/fBQVePzsV7 line 12\.
:&#8203;:\#          got&#8203;: '000000000033'
:&#8203;:\#     expected&#8203;: '030\.031\.032'
:&#8203;:ok 7 \- vector flag \+ custom separator \+ static precision
:&#8203;:not ok 8 \- vector flag \+ custom separator \+ dynamic precision
:&#8203;:\#   Failed test 'vector flag \+ custom separator \+ dynamic precision'
:&#8203;:\#   at /tmp/fBQVePzsV7 line 14\.
:&#8203;:\#          got&#8203;: '000000000033'
:&#8203;:\#     expected&#8203;: '030&#8203;:031&#8203;:032'
:&#8203;:\# Looks like you failed 2 tests of 8\.
:
:Note that we already have tests expecting the current behaviour \- see
:tests 157\-164 in t/op/sprintf\.t\, where the details of test 157 appear
:at the 157th line after the \_\_END\_\_ token\.
:
:That doesn't necessarily mean it's correct though\.

Indeed\, it turns out an earlier patch of mine reordered the parameters
in those tests\. Bad me\. :\(

I have a fix for this on github&#8203;:
&nbsp;  http&#8203;://github\.com/hvds/perl/tree/perl\-83194
&nbsp;  git&#8203;://github\.com/hvds/perl\.git branch perl\-83194
\(but I'm pretty new to git\, so please check carefully\)\.

I'm a little nervous that there may be code out there assuming the buggy
order\,  since the previous change occurred 10 years ago \(Jan 2001\)\.
But given the docs \(added July 2002\) are correct\, I think it's correct
to fix this \- and I expect usage of the particular combination that fails
\(vectorization \+ a dynamic width or precision without an explicit index\)
to be rare to nonexistent outside of testcases\.

Using the fragments specified in the sv\_vcatpvfn\(\) comment 'We allow format
specification elements in this order'\, it should be possible to craft a
regexp to search for examples\, with a bit of difficulty\. I haven't attempted
to do this\.

Hugo

**Comment by p5pRT** [comment]
### From tchrist@perl.com

Hugo wrote&#8203;:

> I'm a little nervous that there may be code out there assuming the buggy
> order\,  since the previous change occurred 10 years ago \(Jan 2001\)\.
> But given the docs \(added July 2002\) are correct\, I think it's correct
> to fix this \- and I expect usage of the particular combination that fails
> \(vectorization \+ a dynamic width or precision without an explicit index\)
> to be rare to nonexistent outside of testcases\.

I do use v quite a bit\, but have certainly been stimied trying
to get dynamic precision to work with it\, like your example of&#8203;:

&nbsp;    printf\("%\*v\.\*X\\n"\, "&#8203;:"\, 3\, "012"\)

Ordering requirements are not apparent\, and apparently not \(well\-?\)
documented\, either\.  The best examples come from the tests\, not the
documentation&#8203;:

&nbsp;    % cd perl\-git
&nbsp;    % tcgrep \-l '\\bs?printf\.\*%\\S\*v' t
&nbsp;    t/comp/use\.t
&nbsp;    t/io/utf8\.t
&nbsp;    t/op/bop\.t
&nbsp;    t/op/length\.t
&nbsp;    t/op/pack\.t
&nbsp;    t/op/sprintf2\.t
&nbsp;    t/op/tr\.t
&nbsp;    t/op/ver\.t
&nbsp;    t/run/switches\.t

&nbsp;    % tcgrep \-c \!&#8203;:2 \`\!\!\` | sort \-t&#8203;: \-k2rn
&nbsp;    tcgrep \-c '\\bs?printf\.\*%\\S\*v' \`tcgrep \-l '\\bs?printf\.\*%\\S\*v' t\` | sort \-t&#8203;: \-k2rn
&nbsp;    t/op/ver\.t&#8203;:35
&nbsp;    t/op/bop\.t&#8203;:11
&nbsp;    t/op/length\.t&#8203;:4
&nbsp;    t/io/utf8\.t&#8203;:3
&nbsp;    t/op/pack\.t&#8203;:3
&nbsp;    t/op/tr\.t&#8203;:3
&nbsp;    t/comp/use\.t&#8203;:1
&nbsp;    t/op/sprintf2\.t&#8203;:1
&nbsp;    t/run/switches\.t&#8203;:1

Speaking of the v flag\, could someone please give me an example of where
using the V flag makes a difference?  Its only example is on line 284 of
t/op/64bitint\.t&#8203;:

&nbsp;    print "not " unless \(sprintf "%\#Vx"\, 1\<\<63\) eq '0x8000000000000000';

I can't figure out any situation where adding the V changes what's returned\.

thanks\,

\-\-tom

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

On Sun\, Feb 6\, 2011 at 4&#8203;:46 PM\, Tom Christiansen \<tchrist@&#8203;perl\.com> wrote&#8203;:
> Speaking of the v flag\, could someone please give me an example of where
> using the V flag makes a difference?  Its only example is on line 284 of
> t/op/64bitint\.t&#8203;:
>
>    print "not " unless \(sprintf "%\#Vx"\, 1\<\<63\) eq '0x8000000000000000';
>
> I can't figure out any situation where adding the V changes what's returned\.

Perlfunc \(for 5\.12\.2\) documents this&#8203;:

&nbsp;    The size specifier "V" has no effect for Perl code\, but is
supported for compatibility with XS code\.  It means "use the standard
size for a Perl integer or floating\-point number"\, which is the
default\.

Ambrus

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Feb 06 07&#8203;:46&#8203;:49 2011\, tom christiansen wrote&#8203;:
> Hugo wrote&#8203;:
> 
> > I'm a little nervous that there may be code out there assuming the
> buggy
> > order\,  since the previous change occurred 10 years ago \(Jan 2001\)\.
> > But given the docs \(added July 2002\) are correct\, I think it's
> correct
> > to fix this \- and I expect usage of the particular combination that
> fails
> > \(vectorization \+ a dynamic width or precision without an explicit
> index\)
> > to be rare to nonexistent outside of testcases\.
> 
> I do use v quite a bit\, but have certainly been stimied trying
> to get dynamic precision to work with it\, like your example of&#8203;:
> 
>     printf\("%\*v\.\*X\\n"\, "&#8203;:"\, 3\, "012"\)
> 
> Ordering requirements are not apparent\, and apparently not \(well\-?\)
> documented\, either\.

Would you recommend that I apply the patch\, or not?

**Comment by p5pRT** [comment]
### From tchrist@perl.com

\<\<\<Hugo>>>

>>> I'm a little nervous that there may be code out there assuming the
>>> buggy order\,  since the previous change occurred 10 years ago \(Jan
>>> 2001\)\. But given the docs \(added July 2002\) are correct\, I think
>>> it's correct to fix this \- and I expect usage of the particular
>>> combination that fails \(vectorization \+ a dynamic width or
>>> precision without an explicit index\) to be rare to nonexistent
>>> outside of testcases\.

\<\<me>>

>> I do use v quite a bit\, but have certainly been stimied trying
>> to get dynamic precision to work with it\, like your example of&#8203;:
>>
>>     printf\("%\*v\.\*X\\n"\, "&#8203;:"\, 3\, "012"\)
>>
>> Ordering requirements are not apparent\, and apparently not \(well\-?\)
>> documented\, either\.

\<spout>

> Would you recommend that I apply the patch\, or not?

Yes\, I \*think\* so\.

I suppose I share Hugo's nervousness\, but I'm not sure of the scope\. Is
the nervousness fear that someone is trying to do this but it's broken\,
and so if we fix it we'll unbreak their broken programs \-\- and so break
them from the standpoint of exact output correspondence? If so\, that
would appear to argue that they're broken in the first place\.

It's not proof or anything\, only evidence\, but I'd perhaps
feel more comfortable if I could run a

&nbsp;    /\\bs?printf\\b\.\*%\\S\*v/

search on all of CPAN to get a feel for what's out there\.

&nbsp;        \(Yes\, one \*can\* craft a better pattern\.\)

Because I tend to be overly conservative in matters of backwards
compatibility\, I may not be the right person to ask about this\.

I will say that \*not\* being able to plug in "dynamic precision"
widths nicely when you use vector prints is annoying\.

At least you can use variable in interpolation for that\,
something you can't do if you want a custom separator;
interpolating the separator is the only way to do it\.

\-\-tom

**Comment by p5pRT** [comment]
### From @hvds

Hugo> I'm a little nervous that there may be code out there assuming the
Hugo> buggy order\,  since the previous change occurred 10 years ago \(Jan
Hugo> 2001\)\. But given the docs \(added July 2002\) are correct\, I think
Hugo> it's correct to fix this \- and I expect usage of the particular
Hugo> combination that fails \(vectorization \+ a dynamic width or
Hugo> precision without an explicit index\) to be rare to nonexistent
Hugo> outside of testcases\.

Tom> I do use v quite a bit\, but have certainly been stimied trying
Tom> to get dynamic precision to work with it\, like your example of&#8203;:
Tom>
Tom>     printf\("%\*v\.\*X\\n"\, "&#8203;:"\, 3\, "012"\)
Tom>
Tom> Ordering requirements are not apparent\, and apparently not \(well\-?\)
Tom> documented\, either\.

See the subsection "order of arguments" in \`perldoc \-f sprintf\`\. That's
"the docs" referred to above\.

spout> Would you recommend that I apply the patch\, or not?

Tom> Yes\, I \*think\* so\.

Tom> I suppose I share Hugo's nervousness\, but I'm not sure of the scope\. Is
Tom> the nervousness fear that someone is trying to do this but it's broken\,
Tom> and so if we fix it we'll unbreak their broken programs \-\- and so break
Tom> them from the standpoint of exact output correspondence? If so\, that
Tom> would appear to argue that they're broken in the first place\.

I would argue that they're broken in the first place\, since the sprintf
docs give the correct argument order\. The only weakness in that position\,
as I see it\, is that the tests underscored the buggy behaviour\.

Tom> It's not proof or anything\, only evidence\, but I'd perhaps
Tom> feel more comfortable if I could run a
Tom>
Tom>     /\\bs?printf\\b\.\*%\\S\*v/
Tom>
Tom> search on all of CPAN to get a feel for what's out there\.
Tom>
Tom>         \(Yes\, one \*can\* craft a better pattern\.\)

At the very least\, if the search is for formats affected by this fix\,
one should look for a \\\* following the v\.

Tom> Because I tend to be overly conservative in matters of backwards
Tom> compatibility\, I may not be the right person to ask about this\.
Tom>
Tom> I will say that \*not\* being able to plug in "dynamic precision"
Tom> widths nicely when you use vector prints is annoying\.

I noticed that issue while looking into this one\, and it also looks
like an easy fix\. I probably won't get to it before the weekend though\.

Hugo

**Comment by p5pRT** [comment]
### From andrew@cleverdomain.org

On Monday\, February 07\, 2011 01&#8203;:00&#8203;:21 AM hv@&#8203;crypt\.org wrote&#8203;:
> Tom> It's not proof or anything\, only evidence\, but I'd perhaps
> Tom> feel more comfortable if I could run a
> Tom>
> Tom>     /\\bs?printf\\b\.\*%\\S\*v/
> Tom>
> Tom> search on all of CPAN to get a feel for what's out there\.
> Tom>
> Tom>         \(Yes\, one \*can\* craft a better pattern\.\)
> 
> At the very least\, if the search is for formats affected by this fix\,
> one should look for a \\\* following the v\.

I did a search on grep\.cpan\.me \(thanks dg for the nifty tool\!\) using a 
modified version of Tom's regex and I don't see any sign of users\. 
http&#8203;://grep\.cpan\.me/?q=%5Cbs%3Fprintf%5Cb\.\*%25%5CS\*v%5CS\*%5C\* returns results\, 
but they're all noise\.

Andrew

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Feb 06 04&#8203;:54&#8203;:57 2011\, hv wrote&#8203;:
> I wrote&#8203;:
> :"l\.mai@&#8203;web\.de \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> :\[\.\.\.\]
> :&#8203;:is sprintf\('%\.3X'\, 11\)\, '00B'\, 'hex\, zero\-padded to 3 places';
> :&#8203;:is sprintf\('%\.\*X'\, 3\, 11\)\, '00B'\, 'dynamic precision';
> :&#8203;:is sprintf\('%vX'\, '012'\)\, '30\.31\.32'\, 'vector flag';
> :&#8203;:is sprintf\('%\*vX'\, '&#8203;:'\, '012'\)\, '30&#8203;:31&#8203;:32'\, 'vector flag \+ custom
> separator';
> :&#8203;:is sprintf\('%v\.3X'\, '012'\)\, '030\.031\.032'\, 'vector flag \+ static
> precision';
> :&#8203;:is sprintf\('%v\.\*X'\, 3\, '012'\)\, '030\.031\.032'\, 'vector flag \+ dynamic
> precision';
> :&#8203;:is sprintf\('%\*v\.3X'\, '&#8203;:'\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+
> custom separator \+ static precision';
> :&#8203;:is sprintf\('%\*v\.\*X'\, '&#8203;:'\, 3\, '012'\)\, '030&#8203;:031&#8203;:032'\, 'vector flag \+
> custom separator \+ dynamic precision';
> :&#8203;:
> :&#8203;:\_\_END\_\_
> :&#8203;:
> :&#8203;:1\.\.8
> :&#8203;:ok 1 \- hex\, zero\-padded to 3 places
> :&#8203;:ok 2 \- dynamic precision
> :&#8203;:ok 3 \- vector flag
> :&#8203;:ok 4 \- vector flag \+ custom separator
> :&#8203;:ok 5 \- vector flag \+ static precision
> :&#8203;:not ok 6 \- vector flag \+ dynamic precision
> :&#8203;:\#   Failed test 'vector flag \+ dynamic precision'
> :&#8203;:\#   at /tmp/fBQVePzsV7 line 12\.
> :&#8203;:\#          got&#8203;: '000000000033'
> :&#8203;:\#     expected&#8203;: '030\.031\.032'
> :&#8203;:ok 7 \- vector flag \+ custom separator \+ static precision
> :&#8203;:not ok 8 \- vector flag \+ custom separator \+ dynamic precision
> :&#8203;:\#   Failed test 'vector flag \+ custom separator \+ dynamic precision'
> :&#8203;:\#   at /tmp/fBQVePzsV7 line 14\.
> :&#8203;:\#          got&#8203;: '000000000033'
> :&#8203;:\#     expected&#8203;: '030&#8203;:031&#8203;:032'
> :&#8203;:\# Looks like you failed 2 tests of 8\.
> :
> :Note that we already have tests expecting the current behaviour \- see
> :tests 157\-164 in t/op/sprintf\.t\, where the details of test 157 appear
> :at the 157th line after the \_\_END\_\_ token\.
> :
> :That doesn't necessarily mean it's correct though\.
> 
> Indeed\, it turns out an earlier patch of mine reordered the parameters
> in those tests\. Bad me\. :\(
> 
> I have a fix for this on github&#8203;:
>   http&#8203;://github\.com/hvds/perl/tree/perl\-83194
>   git&#8203;://github\.com/hvds/perl\.git branch perl\-83194
> \(but I'm pretty new to git\, so please check carefully\)\.

Thank you\. I have just applied this as 2393ee1\, since CPAN apparently
contains no code that would be affected\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
